### PR TITLE
Beam Init: Waterbag Distribution 

### PR DESCRIPTION
**Pull request by qianglbl**
add a function in ImpactX.cpp to generate 1 particle coordinates (x, y, t, px, py, pt) from a Waterbag distribution with zero mean and std=1.

- [x] merge `development` to include fix from #48
- [x] user-facing documentation added in `parameters.rst`

### CODE REVIEW

File: src/particles/elements/Elemlist.H
Comment (ax3l) [inline src/particles/elements/Elemlist.H]: This file is probably added accidentally, we have this already in development :)

File: src/ImpactX.cpp
Comment (ax3l) [inline src/ImpactX.cpp]: That's the preferred name in C++ for the same functionality:
```suggestion
#include <cmath>
```

**Review by cemitch99 — COMMENTED** [review]
The function waterbag looks good to me.

**Comment by qianglbl** [comment]
change "math.h" into "cmath".

File: src/particles/elements/Elemlist.H
Comment (ax3l) [inline src/particles/elements/Elemlist.H]: We can remove this with `git rm src/particles/elements/Elemlist.H` and then `git commit` it.

File: src/ImpactX.cpp
Comment (ax3l) [inline src/ImpactX.cpp]:
```suggestion
          r1 = 2_rt * amrex::Random(engine) - 1.0_prt;
          r2 = 2_rt * amrex::Random(engine) - 1.0_prt;
          r3 = 2_rt * amrex::Random(engine) - 1.0_prt;
          x1 = 2_rt * amrex::Random(engine) - 1.0_prt;
          x2 = 2_rt * amrex::Random(engine) - 1.0_prt;
          x3 = 2_rt * amrex::Random(engine) - 1.0_prt;
          if(r1*r1+r2*r2+r3*r3+x1*x1+x2*x2+x3*x3<1.0_prt) break;
    }
        x = sqrt(8.0_prt)* r1;
        y = sqrt(8.0_prt) * r2;
        t = sqrt(8.0_prt) * r3;
        px = sqrt(8.0_prt) * x1;
        py = sqrt(8.0_prt) * x2;
        pt = sqrt(8.0_prt) * x3;
```

File: src/ImpactX.cpp
Comment (ax3l) [inline src/ImpactX.cpp]:
```suggestion
        using namespace amrex::literals;

        amrex::ParticleReal r1=0.0, r2=0.0, r3=0.0;
```

File: src/ImpactX.cpp
Comment (ax3l) [inline src/ImpactX.cpp]:
```suggestion
        amrex::ParticleReal x1=0.0, x2=0.0, x3=0.0;
```

File: src/ImpactX.cpp
Comment (ax3l) [inline src/ImpactX.cpp]:
```suggestion
        amrex::RandomEngine engine;
```

File: src/ImpactX.cpp
Comment (ax3l) [inline src/ImpactX.cpp]:
```suggestion
          r1 = 2_prt * amrex::Random(engine) - 1.0_prt;
          r2 = 2_prt * amrex::Random(engine) - 1.0_prt;
          r3 = 2_prt * amrex::Random(engine) - 1.0_prt;
          x1 = 2_prt * amrex::Random(engine) - 1.0_prt;
          x2 = 2_prt * amrex::Random(engine) - 1.0_prt;
          x3 = 2_prt * amrex::Random(engine) - 1.0_prt;
          if(r1*r1+r2*r2+r3*r3+x1*x1+x2*x2+x3*x3<1.0_prt) break;
```

File: src/ImpactX.cpp
Comment (ax3l) [inline src/ImpactX.cpp]:
```suggestion
    void ImpactX::Waterbag (amrex::ParticleReal &x,amrex::ParticleReal &y, amrex::ParticleReal &t,
                            amrex::ParticleReal &px, amrex::ParticleReal &py, amrex::ParticleReal &pt)
```

File: src/ImpactX.H
Comment (ax3l) [inline src/ImpactX.H]:
```suggestion
    void Waterbag (amrex::ParticleReal &x, amrex::ParticleReal &y, amrex::ParticleReal &t,
                   amrex::ParticleReal &px, amrex::ParticleReal &py, amrex::ParticleReal &pt);
```

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [inline src/particles/distribution/Waterbag.H]: Just as a general note:
A rejection method will not be very performant on GPU, since we will see a lot of thread divergence here.

If we can express this as a transformation of a uniform or standard distribution (for which be both have functions as well to pull random numbers from), then this would be better.

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [inline src/particles/distribution/Waterbag.H]: Do you like to expose the mean and std deviation as parameters?
This can be done here and stored as member variable (similar to how we expose beamline element parameters, e.g., in the `Quad` class).

Comment (qianglbl) [reply src/particles/distribution/Waterbag.H]: yes.

File: src/particles/distribution/Waterbag.H
Comment (qianglbl) [reply src/particles/distribution/Waterbag.H]: I have to think about this a bit.

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [reply src/particles/distribution/Waterbag.H]: Ok, I'll then wait for you to add those before we merge it in.

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [reply src/particles/distribution/Waterbag.H]: @cemitch99 added on Slack:
> This can be done.  I have some Fortran code for KV-like and waterbag distributions that use a transformation based on 1D uniform sampling--I can share, discuss with Ji.

File: src/particles/ImpactXParticleContainer.H
Comment (ax3l) [inline src/particles/ImpactXParticleContainer.H:90]: In the Doxygen comment above, the new parameters need to be added :)

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [inline src/particles/distribution/Waterbag.H]:
```suggestion
         * @param sigx,sigy,sigt for zero correlation, these are the related
         *                                   RMS sizes (length); ...
         * @param sigpx,sigpy,sigpt RMS momentum
         * @param muxpx,muypy,mutpt correlation length-momentum
```

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [inline src/particles/distribution/Waterbag.H]:
```suggestion
            constexpr amrex::ParticleReal pi = 3.14159265358979_prt;
```

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [inline src/particles/distribution/Waterbag.H]:
```suggestion
            amrex::ParticleReal ln1,norm,u1,u2;
```

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [inline src/particles/distribution/Waterbag.H]:
```suggestion
        amrex::ParticleReal m_sigmaX,m_sigmaY,m_sigmaT; //! related RMS sizes (length)
        amrex::ParticleReal m_sigmaPx,m_sigmaPy,m_sigmaPt; //! RMS momentum
```

**Review by ax3l — APPROVED** [review]
Thank you @qianglbl, looks great to me!

In a follow-up PR, we can then also add the needed control to set particle type, weighting and charge-to-mass ratio based on particle type :)

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [inline src/particles/distribution/Waterbag.H:68]: @cemitch99 suggested on Slack to add a few inline comments of the form
```C++
// now we calculate this... then that
```

here, to make the logic here easier to follow.

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [inline src/particles/distribution/Waterbag.H:27]: Did we mean to fill the `...` with more info?
```suggestion
         * @param sigx,sigy,sigt for zero correlation, these are the related
         *                       RMS sizes (length)
```

Comment (ax3l) [reply src/particles/distribution/Waterbag.H:27]: @cemitch99 @qianglbl this slipped through, please fix in a follow-up PR :)

File: src/particles/distribution/Waterbag.H
Comment (ax3l) [reply src/particles/distribution/Waterbag.H:68]: @cemitch99 @qianglbl please add this as needed in a follow-up PR :)

**Pull request merged**
